test/e2e/framework/helper/certificates.go

```diff
@@ -20,83 +20,151 @@ import (
 	"context"
 	"crypto/x509"
 	"fmt"
+	"os"
 	"sort"
 	"time"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
 	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	v1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
 	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	clientset "github.com/jetstack/cert-manager/pkg/client/clientset/versioned/typed/certmanager/v1"
 	"github.com/jetstack/cert-manager/test/e2e/framework/log"
-	e2eutil "github.com/jetstack/cert-manager/test/e2e/util"
 )
 
-func (h *Helper) handleResult(ns, name string, cert *cmapi.Certificate, state string, err error) (*cmapi.Certificate, error) {
-	if err != nil {
-		log.Logf("Error waiting for Certificate to become %s: %v", state, err)
-		h.Kubectl(ns).DescribeResource("certificate", name)
-		h.Kubectl(ns).Describe("order", "challenge")
-		h.describeCertificateRequestFromCertificate(ns, cert)
-	}
-	return cert, err
-}
+func (h *Helper) waitPollImmediateCertificate(client clientset.CertificateInterface, name string, check func(*v1.Certificate) bool, interval time.Duration, timeout time.Duration) (*cmapi.Certificate, error) {
+	var certificate *v1.Certificate = nil
+	pollErr := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		var err error
+		certificate, err = client.Get(context.TODO(), name, metav1.GetOptions{})
+		if nil != err {
+			certificate = nil
+			return false, fmt.Errorf("error getting Certificate %v: %v", name, err)
```

Comment:
The `f.CRClient.Create(ctx, testCertificate)` function call that might precede the `waitForMatchingNamedCertificate` call will already wait for the certificated to be created. So it would be unexpected if the certificate was not found I think.

Reply:
I think the PollImmediate should exit if there is an unexpected error. If that is not the case it should continue until there is a timeout.

Reply:
I just noticed there is also a `WaitForCertificateToExist` function that is used, maybe its possible to combine these two calls?